### PR TITLE
dev/event#23 Don't count non-counted participants on Confirm Registration screen

### PR DESCRIPTION
**Pull request by MegaphoneJon**
on screen

Overview
----------------------------------------
If you have an event with a waitlist, and some participant roles are counted (e.g. attendee) and some are not (e.g. host), someone attempting to confirm that they're coming off the waitlist will be unable to, because that screen counts all participants, whether they're countable or not.

Detailed replicated instructions are on the ticket: https://lab.civicrm.org/dev/event/issues/23
As side effect, this also fixes [dev/event#7](https://lab.civicrm.org/dev/event/issues/7), where the same screen counts deleted contacts if they're registered for the event.

Before
----------------------------------------
Unable to come off the waitlist if the total participants, *including non-counted participants*, exceed the max participants.

After
----------------------------------------
Only participants with "counted" roles are counted.
 
Technical Details
----------------------------------------
I ripped out `CRM_Event_BAO_Participant::pendingToConfirmSpaces()`, which calculates the participant count in a faulty way (and is untested) in favor of `CRM_Event_BAO_Participant::eventFull()`, which counts correctly, is tested, and is used everywhere else in the codebase (`pendingToConfirmSpaces()` was only used once).

### CODE REVIEW

File: CRM/Event/BAO/Participant.php [hunk 1]
```diff
@@ -354,18 +361,16 @@ public static function eventFull(
 
     $participantRoles = CRM_Event_PseudoConstant::participantRole(NULL, 'filter = 1');
     $countedStatuses = CRM_Event_PseudoConstant::participantStatus(NULL, 'is_counted = 1');
+    $positiveStatuses = CRM_Event_PseudoConstant::participantStatus(NULL, "class = 'Positive'");
     $waitingStatuses = CRM_Event_PseudoConstant::participantStatus(NULL, "class = 'Waiting'");
     $onWaitlistStatusId = array_search('On waitlist', $waitingStatuses);
 
-    //when we do require only waiting count don't consider counted.
```
Comment (MegaphoneJon) [inline CRM/Event/BAO/Participant.php]: This section here is unrelated to the rest of the PR; `$allStatusIds` is unused in the method so this is just a bit of cleanup.

**Comment by mlutfy** [comment]
It might be unlikely, but some customizations may be calling `pendingToConfirmSpaces()`.

Would it be possible to convert it to a wrapper over `eventFull`, flag it as deprecated (code comment `@deprecated` and add a log message that it will disappear soon?

```
CRM_Core_Error::deprecatedFunctionWarning(...)
```

**Comment by MegaphoneJon** [comment]
Given how incredibly buggy this function is I'd be a bit shocked if someone was successfully using this in an extension, but if that's what it takes to get this merged, sure, no problem.

**Comment by MegaphoneJon** [comment]
@bgm I've incorporated your feedback, thanks for looking at this.

**Comment by mlutfy** [comment]
Test fail is unrelated. Seems like the test node ran out of disk.

**Comment by mlutfy** [comment]
jenkins, test this please

**Comment by mlutfy** [comment]
Looks good to me. I did a bit of poking around, and also confirmed that the function is called only in one place, and did not have tests. Thanks @MegaphoneJon !

**Comment by seamuslee001** [comment]
Test fail unrelated

**Comment by eileenmcnaughton** [comment]
@MegaphoneJon I'm revisiting this code & trying to understand why participantConfirm checks for 'Postive' statuses but all the rest consider is_counted when determining if there is spaces (all the rest includes both api)

The Participant status screen says

image


My suspicion is that it was legacy behaviour at the time when you created this PR and you kept it  because altering it was out of scope - but there might be a reason for the behaviour?

**Comment by seamuslee001** [comment]
@eileenmcnaughton I'm guessing the difference might be in regards to the Pending from waitlist as they wouldn't be "counted" I suspect but might still be positive?

**Comment by seamuslee001** [comment]
actually Its the other way around it seems they are Counted but are not a "positive" status

image

**Comment by eileenmcnaughton** [comment]
@seamuslee001 so I think the idea is that if you are confirming then we are using a narrower definition of 'taken spaces' - which I guess makes sense - I am thinking maybe in addition to supporting `available_spaces` as a calculated field it should support `avaiable_confirm_spaces` or similar?
